### PR TITLE
Reorder class and plan creation; test plan conflict handling

### PR DESCRIPTION
**Pull request by nilebox**
1. Create classes first, plans after that. Plans are referencing parent classes, and the current order is weird (plan is referencing a class which doesn't yet exist in Service Catalog). Also it doesn't allow for implementing https://github.com/kubernetes-incubator/service-catalog/issues/1432
2. Add test for correctly handling plan conflict.

### CODE REVIEW

File: pkg/controller/controller_broker.go [hunk 1]
```diff
@@ -299,44 +299,44 @@ func (c *controller) reconcileClusterServiceBroker(broker *v1beta1.ClusterServic
 		existingServiceClassMap := convertServiceClassListToMap(existingServiceClasses)
 		existingServicePlanMap := convertServicePlanListToMap(existingServicePlans)
 
-		// reconcile the plans that were part of the broker's catalog payload
```
Comment (nilebox) [inline pkg/controller/controller_broker.go]: The diff is misleading. What I have actually done is just moved the "serviceclasses" block before the "plans" block. No other code change.

File: pkg/controller/controller_broker_test.go [hunk 2]
```diff
@@ -358,15 +358,54 @@ func TestReconcileClusterServiceBrokerExistingClusterServiceClassDifferentBroker
...
+	expectedEvent := corev1.EventTypeWarning + " " + errorSyncingCatalogReason + ` Error reconciling ClusterServiceClass (K8S: "SCGUID" ExternalName: "test-serviceclass") (broker "test-broker"): ClusterServiceBroker "test-broker": ClusterServiceClass (K8S: "SCGUID" ExternalName: "test-serviceclass") already exists for Broker "notTheSame"`
+	if e, a := expectedEvent, events[0]; e != a {
+		t.Fatalf("Received unexpected event; expected\n%v, got\n%v", e, a)
+	}
+}
+
+// TestReconcileClusterServiceBrokerExistingClusterServicePlanDifferentClass simulates catalog
+// refresh where broker lists a service plan which matches an existing, already
+// cataloged service plan but the plan points to a different ClusterServiceClass.  Results in an error.
+func TestReconcileClusterServiceBrokerExistingClusterServicePlanDifferentClass(t *testing.T) {
```
Comment (nilebox) [inline pkg/controller/controller_broker_test.go]: Again, the diff is misleading. I have added a new test method for handling plan conflicts. The other tests remain unchanged apart from reordering actions assertion (classes actions go before plans now).

File: pkg/controller/controller_broker_test.go [hunk 3]
```diff
@@ -358,15 +358,54 @@ func TestReconcileClusterServiceBrokerExistingClusterServiceClassDifferentBroker
...
+	assertClusterServiceBrokerReadyFalse(t, updatedClusterServiceBroker)
+
+	// verify no kube resources created
+	kubeActions := fakeKubeClient.Actions()
+	assertNumberOfActions(t, kubeActions, 0)
+
+	events := getRecordedEvents(testController)
+	assertNumEvents(t, events, 1)
+
+	expectedEvent := corev1.EventTypeWarning + " " + errorSyncingCatalogReason + ` Error reconciling ClusterServiceClass (K8S: "SCGUID" ExternalName: "test-serviceclass") (broker "test-broker"): ClusterServiceBroker "test-broker": ClusterServiceClass (K8S: "SCGUID" ExternalName: "test-serviceclass") already exists for Broker "notTheSame"`
```
Comment (nilebox) [inline pkg/controller/controller_broker_test.go]: The classes conflict test was actually checking that we fail to create plan before, which was weird.
With new order it checks that Service Catalog fails to create a class with duplicate name - 👌

File: pkg/controller/controller_broker_test.go [hunk 4]
```diff
@@ -358,15 +358,54 @@ func TestReconcileClusterServiceBrokerExistingClusterServiceClassDifferentBroker
...
+	assertClusterServiceBrokerReadyFalse(t, updatedClusterServiceBroker)
+
+	// verify no kube resources created
+	kubeActions := fakeKubeClient.Actions()
+	assertNumberOfActions(t, kubeActions, 0)
+
+	events := getRecordedEvents(testController)
+	assertNumEvents(t, events, 1)
+
+	expectedEvent := corev1.EventTypeWarning + " " + errorSyncingCatalogReason + ` Error reconciling ClusterServiceClass (K8S: "SCGUID" ExternalName: "test-serviceclass") (broker "test-broker"): ClusterServiceClass (K8S: "SCGUID" ExternalName: "test-serviceclass") already exists for Broker "notTheSame"`
```
Comment (nilebox) [inline pkg/controller/controller_broker_test.go]: The `TestReconcileClusterServiceBrokerExistingClusterServiceClassDifferentBroker` (conflict with duplicate class pointing a different broker) was actually checking for failure on plan creation.
With the new order it checks that we fail to create a class with duplicate name.

File: pkg/controller/controller_broker_test.go (hunk 2 above)
Comment (pmorie) [reply pkg/controller/controller_broker_test.go]: I find the diff-assessment behavior of git to be highly annoying in this area

**Pull request merged**